operators/src/pro/cache/cache_tiles.rs

```diff
@@ -314,23 +309,354 @@ impl<T: Pixel> Stream for CacheTileStream<T> {
+        match self {
+            Self::Empty(empty_grid) => empty_grid.heap_byte_size(),
+            Self::Compressed(compressed_grid) => compressed_grid.heap_byte_size(),
+        }
+    }
+}
+
+impl<D, T> ByteSize for CompressedMaskedGrid<D, T> {
+    fn heap_byte_size(&self) -> usize {
+        self.data.heap_byte_size() + self.mask.heap_byte_size()
+    }
+}
+
+pub trait CompressedRasterTileExt<D, T>
+where
+    Self: Sized,
+{
+    fn compress_tile<F>(tile: RasterTile<D, T>, compression: F) -> Result<Self, CacheError>
+    where
+        F: Fn(&[u8]) -> Vec<u8>;
+
+    fn decompress_tile<F>(&self, decompression: F) -> Result<RasterTile<D, T>, CacheError>
+    where
+        F: Fn(&[u8]) -> Result<Vec<u8>, CacheError>;
+
+    fn compressed_data_size(&self) -> usize;
+}
+
+impl<T> CompressedRasterTileExt<GridShape2D, T> for BaseTile<CompressedGridOrEmpty<GridShape2D, T>>
+where
+    T: Copy,
+{
+    fn compress_tile<F>(
+        tile: RasterTile<GridShape2D, T>,
+        compression: F,
+    ) -> Result<Self, CacheError>
+    where
+        F: Fn(&[u8]) -> Vec<u8>,
+    {
+        let compressed_grid = CompressedGridOrEmpty::compress_grid(&tile.grid_array, compression);
+        Ok(Self {
+            grid_array: compressed_grid,
+            time: tile.time,
+            cache_hint: tile.cache_hint,
+            global_geo_transform: tile.global_geo_transform,
+            properties: tile.properties,
+            tile_position: tile.tile_position,
+        })
+    }
+
+    fn decompress_tile<F>(&self, decompression: F) -> Result<RasterTile<GridShape2D, T>, CacheError>
+    where
+        F: Fn(&[u8]) -> Result<Vec<u8>, CacheError>,
+    {
+        let grid_array = match &self.grid_array {
+            CompressedGridOrEmpty::Empty(empty_grid) => GridOrEmpty::Empty(*empty_grid),
+            CompressedGridOrEmpty::Compressed(compressed_grid) => {
+                let decompressed_grid = compressed_grid.decompress_masked_grid(decompression)?;
+                GridOrEmpty::Grid(decompressed_grid)
+            }
+        };
+
+        Ok(RasterTile {
+            grid_array,
+            time: self.time,
+            cache_hint: self.cache_hint,
+            global_geo_transform: self.global_geo_transform,
+            properties: self.properties.clone(),
+            tile_position: self.tile_position,
+        })
+    }
+
+    fn compressed_data_size(&self) -> usize {
+        self.grid_array.compressed_data_size()
+    }
+}
+
+impl<T> CacheElement for RasterTile2D<T>
+where
+    T: Pixel + CacheElementSubType,
+    CompressedRasterTile2D<T>:
+        CompressedRasterTileExt<GridShape2D, T> + CacheBackendElement<Query = RasterQueryRectangle>,
+{
+    type StoredCacheElement = CompressedRasterTile2D<T>;
+    type Query = RasterQueryRectangle;
+    type ResultStream = CacheTileStream<T>;
+
+    fn into_stored_element(self) -> Self::StoredCacheElement {
+        CompressedRasterTile2D::compress_tile(self, lz4_flex::compress_prepend_size)
```

Comment:
why prepend the size? We know it either way, as we store the compressed data in a vec with known length?

Reply:
[4e25ba7](https://github.com/geo-engine/geoengine/pull/848/commits/4e25ba786092b3badcad6b52e91bb75bb020e586)